File: src/oauth/discovery.ts
// OAuth endpoint discovery.
//
// Resolves the authorization server's endpoints in two hops:
//   1. RFC 9728 — GET <ADAPTER_BASE_URL>/.well-known/oauth-protected-resource
//      → authorization_servers[0] gives the AS base.
//   2. RFC 8414 — GET <AS base>/.well-known/oauth-authorization-server
//      → authorization_endpoint, token_endpoint, registration_endpoint?,
//        dpop_signing_alg_values_supported?
//
// OKTA_ISSUER override: the adapter enforces cnf.jkt equality, so the DPoP-bound
// token must be minted by the actual issuer (Okta). When OKTA_ISSUER is set we
// ALSO fetch <OKTA_ISSUER>/.well-known/openid-configuration and use ITS
// authorize/token endpoints — binding DPoP directly at Okta. Without the
// override we rely on the adapter's token endpoint forwarding the DPoP proof +
// nonce to Okta unchanged.
//
// Resolved metadata is cached in memory for the process lifetime.
import type { Config } from "../config.ts";
import { logger as defaultLogger, type Logger } from "../logger.ts";

export interface Endpoints {
  issuer: string;
  authorizationEndpoint: string;
  tokenEndpoint: string;
  registrationEndpoint?: string;
  dpopAlgs?: string[];
}

export interface DiscoveryDeps {
  /** Injectable for tests; defaults to the global fetch. */
  fetch?: typeof fetch;
  logger?: Logger;
}

interface ProtectedResourceMetadata {
  authorization_servers?: string[];
}

interface AuthServerMetadata {
  issuer?: string;
  authorization_endpoint?: string;
  token_endpoint?: string;
  registration_endpoint?: string;
  dpop_signing_alg_values_supported?: string[];
}

const cache = new Map<string, Endpoints>();

/** Clear the in-memory discovery cache (used by tests). */
export function clearDiscoveryCache(): void {
  cache.clear();
}

function stripTrailingSlash(url: string): string {
  return url.replace(/\/+$/, "");
}

function hostOf(url: string): string {
  try {
    return new URL(url).host;
  } catch {
    return "?";
  }
}

async function getJson<T>(
  doFetch: typeof fetch,
  url: string,
  what: string,
): Promise<T> {
  let res: Response;
  try {
    res = await doFetch(url, {
      headers: { Accept: "application/json" },
    });
  } catch (cause) {
    throw new Error(`Failed to fetch ${what} (${url}): ${String(cause)}`);
  }
  if (!res.ok) {
    throw new Error(`Failed to fetch ${what} (${url}): HTTP ${res.status}`);
  }
  return (await res.json()) as T;
}

/**
 * Resolve the OAuth endpoints for the configured adapter, honoring the optional
 * OKTA_ISSUER override. Results are cached per (ADAPTER_BASE_URL, OKTA_ISSUER).
 */
export async function resolveEndpoints(
  config: Config,
  deps: DiscoveryDeps = {},
): Promise<Endpoints> {
  const doFetch = deps.fetch ?? fetch;
  const logger = deps.logger ?? defaultLogger;

  const cacheKey = `${config.ADAPTER_BASE_URL}|${config.OKTA_ISSUER ?? ""}`;
  const cached = cache.get(cacheKey);
  if (cached) return cached;

  // 1. RFC 9728 — protected-resource metadata → authorization server base.
  const base = stripTrailingSlash(config.ADAPTER_BASE_URL);
  const prm = await getJson<ProtectedResourceMetadata>(
    doFetch,
    `${base}/.well-known/oauth-protected-resource`,
    "protected-resource metadata",
  );
  const asBase = prm.authorization_servers?.[0];
  if (!asBase) {
    throw new Error(
      "protected-resource metadata has no authorization_servers[0]",
    );
  }

  // 2. RFC 8414 — authorization-server metadata.
  const asMeta = await getJson<AuthServerMetadata>(
    doFetch,
    `${stripTrailingSlash(asBase)}/.well-known/oauth-authorization-server`,
    "authorization-server metadata",
  );
  if (!asMeta.authorization_endpoint || !asMeta.token_endpoint) {
    throw new Error(
      "authorization-server metadata missing authorization_endpoint/token_endpoint",
    );
  }

  let endpoints: Endpoints = {
    issuer: asMeta.issuer ?? stripTrailingSlash(asBase),
    authorizationEndpoint: asMeta.authorization_endpoint,
    tokenEndpoint: asMeta.token_endpoint,
    registrationEndpoint: asMeta.registration_endpoint,
    dpopAlgs: asMeta.dpop_signing_alg_values_supported,
  };

  // OKTA_ISSUER override: mint the DPoP-bound token at Okta directly.
  if (config.OKTA_ISSUER) {
    const oidc = await getJson<AuthServerMetadata>(
      doFetch,
      `${stripTrailingSlash(config.OKTA_ISSUER)}/.well-known/openid-configuration`,
      "Okta openid-configuration",
    );
    if (!oidc.authorization_endpoint || !oidc.token_endpoint) {
      throw new Error(
        "Okta openid-configuration missing authorization_endpoint/token_endpoint",
      );
    }
    endpoints = {
      issuer: oidc.issuer ?? stripTrailingSlash(config.OKTA_ISSUER),
      authorizationEndpoint: oidc.authorization_endpoint,
      tokenEndpoint: oidc.token_endpoint,
      registrationEndpoint:
        oidc.registration_endpoint ?? endpoints.registrationEndpoint,
      dpopAlgs:
        oidc.dpop_signing_alg_values_supported ?? endpoints.dpopAlgs,
    };
  }

  if (
    endpoints.dpopAlgs &&
    endpoints.dpopAlgs.length > 0 &&
    !endpoints.dpopAlgs.includes(config.DPOP_ALG)
  ) {
    logger.warn("oauth.dpop.alg_unsupported", {
      configured: config.DPOP_ALG,
      supported: endpoints.dpopAlgs,
    });
  }

  logger.info("oauth.discovery.resolved", {
    issuer_host: hostOf(endpoints.issuer),
    authorize_host: hostOf(endpoints.authorizationEndpoint),
    token_host: hostOf(endpoints.tokenEndpoint),
    override: Boolean(config.OKTA_ISSUER),
  });

  cache.set(cacheKey, endpoints);
  return endpoints;
}

File: internal/oauth/discovery.go
// Package oauth implements OAuth endpoint discovery, the PKCE auth-code flow
// over a loopback redirect, and the DPoP token client. Go port of src/oauth/*.ts.
package oauth

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"sync"

	"github.com/joewitt99/bridge-mcp-client/internal/config"
	"github.com/joewitt99/bridge-mcp-client/internal/logx"
)

// Doer is the slice of *http.Client the oauth package needs (eases testing).
type Doer interface {
	Do(*http.Request) (*http.Response, error)
}

// Endpoints are the resolved OAuth endpoints.
type Endpoints struct {
	Issuer                string
	AuthorizationEndpoint string
	TokenEndpoint         string
	RegistrationEndpoint  string
	DpopAlgs              []string
}

var (
	discoveryCacheMu sync.Mutex
	discoveryCache   = map[string]Endpoints{}
)

// ClearDiscoveryCache empties the in-memory cache (used by tests).
func ClearDiscoveryCache() {
	discoveryCacheMu.Lock()
	discoveryCache = map[string]Endpoints{}
	discoveryCacheMu.Unlock()
}

type protectedResourceMeta struct {
	AuthorizationServers []string `json:"authorization_servers"`
}

type authServerMeta struct {
	Issuer                        string   `json:"issuer"`
	AuthorizationEndpoint         string   `json:"authorization_endpoint"`
	TokenEndpoint                 string   `json:"token_endpoint"`
	RegistrationEndpoint          string   `json:"registration_endpoint"`
	DpopSigningAlgValuesSupported []string `json:"dpop_signing_alg_values_supported"`
}

// ResolveEndpoints resolves the OAuth endpoints via RFC 9728 → RFC 8414, with
// the optional OKTA_ISSUER override (openid-configuration). Results are cached
// per (ADAPTER_BASE_URL, OKTA_ISSUER) for the process lifetime.
func ResolveEndpoints(ctx context.Context, cfg config.Config, doer Doer, logger *logx.Logger) (Endpoints, error) {
	if logger == nil {
		logger = logx.Default
	}
	cacheKey := cfg.AdapterBaseURL + "|" + cfg.OktaIssuer
	discoveryCacheMu.Lock()
	if ep, ok := discoveryCache[cacheKey]; ok {
		discoveryCacheMu.Unlock()
		return ep, nil
	}
	discoveryCacheMu.Unlock()

	base := strings.TrimRight(cfg.AdapterBaseURL, "/")
	var prm protectedResourceMeta
	if err := getJSON(ctx, doer, base+"/.well-known/oauth-protected-resource", "protected-resource metadata", &prm); err != nil {
		return Endpoints{}, err
	}
	if len(prm.AuthorizationServers) == 0 {
		return Endpoints{}, fmt.Errorf("protected-resource metadata has no authorization_servers[0]")
	}
	asBase := strings.TrimRight(prm.AuthorizationServers[0], "/")

	var as authServerMeta
	if err := getJSON(ctx, doer, asBase+"/.well-known/oauth-authorization-server", "authorization-server metadata", &as); err != nil {
		return Endpoints{}, err
	}
	if as.AuthorizationEndpoint == "" || as.TokenEndpoint == "" {
		return Endpoints{}, fmt.Errorf("authorization-server metadata missing authorization_endpoint/token_endpoint")
	}

	ep := Endpoints{
		Issuer:                orDefault(as.Issuer, asBase),
		AuthorizationEndpoint: as.AuthorizationEndpoint,
		TokenEndpoint:         as.TokenEndpoint,
		RegistrationEndpoint:  as.RegistrationEndpoint,
		DpopAlgs:              as.DpopSigningAlgValuesSupported,
	}

	if cfg.OktaIssuer != "" {
		var oidc authServerMeta
		issuer := strings.TrimRight(cfg.OktaIssuer, "/")
		if err := getJSON(ctx, doer, issuer+"/.well-known/openid-configuration", "Okta openid-configuration", &oidc); err != nil {
			return Endpoints{}, err
		}
		if oidc.AuthorizationEndpoint == "" || oidc.TokenEndpoint == "" {
			return Endpoints{}, fmt.Errorf("Okta openid-configuration missing authorization_endpoint/token_endpoint")
		}
		ep.Issuer = orDefault(oidc.Issuer, issuer)
		ep.AuthorizationEndpoint = oidc.AuthorizationEndpoint
		ep.TokenEndpoint = oidc.TokenEndpoint
		if oidc.RegistrationEndpoint != "" {
			ep.RegistrationEndpoint = oidc.RegistrationEndpoint
		}
		if len(oidc.DpopSigningAlgValuesSupported) > 0 {
			ep.DpopAlgs = oidc.DpopSigningAlgValuesSupported
		}
	}

	if len(ep.DpopAlgs) > 0 && !contains(ep.DpopAlgs, cfg.DpopAlg) {
		logger.Warn("oauth.dpop.alg_unsupported", logx.Fields{
			"configured": cfg.DpopAlg,
			"supported":  ep.DpopAlgs,
		})
	}

	logger.Info("oauth.discovery.resolved", logx.Fields{
		"issuer_host":    hostOf(ep.Issuer),
		"authorize_host": hostOf(ep.AuthorizationEndpoint),
		"token_host":     hostOf(ep.TokenEndpoint),
		"override":       cfg.OktaIssuer != "",
	})

	discoveryCacheMu.Lock()
	discoveryCache[cacheKey] = ep
	discoveryCacheMu.Unlock()
	return ep, nil
}

func getJSON(ctx context.Context, doer Doer, rawURL, what string, dst any) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, rawURL, nil)
	if err != nil {
		return err
	}
	req.Header.Set("Accept", "application/json")
	res, err := doer.Do(req)
	if err != nil {
		return fmt.Errorf("failed to fetch %s (%s): %w", what, rawURL, err)
	}
	defer res.Body.Close()
	if res.StatusCode < 200 || res.StatusCode >= 300 {
		return fmt.Errorf("failed to fetch %s (%s): HTTP %d", what, rawURL, res.StatusCode)
	}
	if err := json.NewDecoder(res.Body).Decode(dst); err != nil {
		return fmt.Errorf("failed to decode %s: %w", what, err)
	}
	return nil
}

func orDefault(v, def string) string {
	if v == "" {
		return def
	}
	return v
}

func contains(ss []string, s string) bool {
	for _, x := range ss {
		if x == s {
			return true
		}
	}
	return false
}

func hostOf(rawURL string) string {
	u, err := url.Parse(rawURL)
	if err != nil || u.Host == "" {
		return "?"
	}
	return u.Host
}
